### PR TITLE
Fix and update type hints for `make_functional.py`

### PR DESCRIPTION
**Pull request by XuehaiPan**
Changes in details:

- Fix and update some out-of-date type hints in `_functorch/make_functional.py`.
- ~Explicitly use `OrderedDict` for order-sensitive mappings.~

	In `create_names_map()`, `_swap_state()`, and `FunctionalModuleWithBuffers.__init__()`, the unordered `dict` was used. The key order should be preserved for `dict.items()` while it is required to `zip` with a tuple of `params`/`buffers`. Although since Python 3.6, the built-in dictionary is insertion ordered ([PEP 468](https://peps.python.org/pep-0468)). Explicit is better than implicit.

cc @zou3519

### CODE REVIEW

File: torch/_functorch/make_functional.py
Comment (zou3519) [inline torch/_functorch/make_functional.py:21]: Is it possible to add this file to lintrunner.toml so we can check in CI that the type annotations are correct? https://github.com/pytorch/pytorch/blob/master/.lintrunner.toml

File: torch/_functorch/make_functional.py
Comment (zou3519) [inline torch/_functorch/make_functional.py:86]: I don't buy the dict -> OrderedDict change:
- Performance. OrderedDict has a different implementation than dict and there isn't a need to construct a new OrderedDict around a dict here.
- As you mentioned, dict() is already ordered. As of Python 3.7, the language standard says that dict() is ordered. PyTorch does not support Python < 3.7, so there are no BC concerns.
- consistency: I'd like to see buy in for "prefer OrderedDict to dict" in other parts of the PyTorch codebase.

**Review by zou3519 — COMMENTED** [review]
Thanks for the PR, @XuehaiPan. We should try to enable mypy checking for this file in PyTorch CI (to check the type hints and prevent regressions -- thanks for catching incorrect type annotations).

I don't buy using OrderedDict over dict here -- if you feel strongly about it, could we please leave it to a separate discussion?

File: torch/_functorch/make_functional.py
Comment (XuehaiPan) [reply torch/_functorch/make_functional.py:86]: > I don't buy the dict -> OrderedDict change:
> 
> * Performance. OrderedDict has a different implementation than dict and there isn't a need to construct a new OrderedDict around a dict here.
> * As you mentioned, dict() is already ordered. As of Python 3.7, the language standard says that dict() is ordered. PyTorch does not support Python < 3.7, so there are no BC concerns.

Let's pass on this since [the language standard guarantees the insertion order for `dict` since Python 3.7](https://docs.python.org/3/library/collections.html#ordereddict-objects). The builtin `dict` is faster for iteration.

```python
In [1]: from collections import OrderedDict

In [2]: iterable = tuple((str(i), i) for i in range(100))

In [3]: %timeit dict(iterable)
2.37 µs ± 20.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [4]: %timeit OrderedDict(iterable)
8.42 µs ± 86.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [5]: d = dict(iterable)

In [6]: od = OrderedDict(iterable)

In [7]: %timeit list(d.items())
2.41 µs ± 83.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [8]: %timeit list(od.items())
4.05 µs ± 87.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

> * consistency: I'd like to see buy in for "prefer OrderedDict to dict" in other parts of the PyTorch codebase.

https://github.com/pytorch/pytorch/blob/3120054c151d37fdf43963dbb60ab420908f48cf/torch/nn/modules/module.py#L448-L463

The attributes in `nn.Module`, such as `_paramters`, `_buffers`, `_modules` are stored as `OrderedDict`.

Note that, `del` and `set` attribute in `_swap_state` changes the order of items in `_paramters` and `_buffers`. (`del + set` is equivalent to `move_to_end()`)

https://github.com/pytorch/pytorch/blob/3120054c151d37fdf43963dbb60ab420908f48cf/torch/_functorch/make_functional.py#L134-L142

File: torch/_functorch/make_functional.py
Comment (XuehaiPan) [reply torch/_functorch/make_functional.py:20]: Done.

**Review by zou3519 — APPROVED** [review]
thank you!

**Comment by XuehaiPan** [comment]
@zou3519 Hi, any update for this PR? I opened a follow-up PR in #92218 which may share some relevant changes.

**Comment by kshitij12345** [comment]
Once approved by a PyTorch maintainer, you can trigger the merge yourself by commenting `@pytorchbot merge`. 

More ref about the pytorchbot: https://github.com/pytorch/pytorch/wiki/Bot-commands

**Pull request closed without merging**
